scriptworker/cot/verify.py

```diff
@@ -714,64 +741,109 @@ def _sort_dependencies_by_name_then_task_id(dependencies):
 
 
 # download_cot_artifacts {{{1
-async def download_cot_artifacts(chain, artifact_dict):
-    """Call ``download_cot_artifact`` in parallel for each key/value in ``artifact_dict``.
+async def download_cot_artifacts(chain):
+    """Call ``download_cot_artifact`` in parallel for each "upstreamArtifacts".
+
+    Optional artifacts are allowed to not be downloaded.
 
     Args:
         chain (ChainOfTrust): the chain of trust object
-        artifact_dict (dict): maps task_id to list of paths of artifacts to download
 
     Returns:
-        list: list of full paths to artifacts
+        list: list of full paths to downloaded artifacts. Failed optional artifacts
+        aren't returned
 
     Raises:
-        CoTError: on chain of trust sha validation error
-        DownloadError: on download error
+        CoTError: on chain of trust sha validation error, on a mandatory artifact
+        DownloadError: on download error on a mandatory artifact
 
     """
-    tasks = []
-    for task_id, paths in artifact_dict.items():
+    upstream_artifacts = chain.task['payload'].get('upstreamArtifacts', [])
+    all_artifacts_per_task_id = get_all_artifacts_per_task_id(chain, upstream_artifacts)
+
+    mandatory_artifact_tasks = []
+    optional_artifact_tasks = []
+    for task_id, paths in all_artifacts_per_task_id.items():
         for path in paths:
-            tasks.append(
-                asyncio.ensure_future(
-                    download_cot_artifact(
-                        chain, task_id, path
-                    )
-                )
-            )
-    full_paths = await raise_future_exceptions(tasks)
-    return full_paths
+            coroutine = asyncio.ensure_future(download_cot_artifact(chain, task_id, path))
+
+            if is_artifact_optional(chain, task_id, path):
+                optional_artifact_tasks.append(coroutine)
+            else:
+                mandatory_artifact_tasks.append(coroutine)
+
+    mandatory_artifacts_paths = await raise_future_exceptions(mandatory_artifact_tasks)
+    succeeded_optional_artifacts_paths, failed_optional_artifacts = \
+        await get_results_and_future_exceptions(optional_artifact_tasks)
 
+    if failed_optional_artifacts:
+        log.warn('Could not download {} artifacts: {}'.format(len(failed_optional_artifacts), failed_optional_artifacts))
 
-# download_firefox_cot_artifacts {{{1
-async def download_firefox_cot_artifacts(chain):
-    """Download artifacts needed for firefox chain of trust verification.
+    return mandatory_artifacts_paths + succeeded_optional_artifacts_paths
 
-    This is only task-graph.json so far.
+
+def is_task_required_by_any_mandatory_artifact(chain, task_id):
```

Comment:
There's a lot of new artifact functions in `cot.verify`. I'm wondering if they belong here or `artifacts.py`, but I don't have a strong opinion here. If we're going to move functions around, though, we already have a major version bump coming :)

Reply:
I'm glad we think alike. I started to put some of these functions in `artifacts.py`. However, they have `chain` as an argument, which makes me think it should somewhere under `cot`. I can spin off a new package like `cot.artifacts` in a quick follow up, if you're interested.

Reply:
Other than the docstring I'm ok leaving the other review comments alone, unless you think we should act on them.

Reply:
Docstring was fixed in https://github.com/mozilla-releng/scriptworker/pull/169/commits/4d448b080a4942eb7f9c43dc4c4c7ec0ad1059b8 . Let's handle the rest in followups, then